### ISSUE TITLE
[valid-expect] support type validation when using `expect` generics

### ISSUE THREAD

**Issue by moshest** [jest-community/eslint-plugin-jest#1299]
On some of my tests I use Jest to test my type definitions and validate the types behave as I expected.

For example:

```ts
type Foo = { bar: number };

expect<Foo>({ bar: 123 });

expect<Foo>({ 
  // @ts-expect-error `bar` should be number
  bar: 'string',
});
```

However, both lines get errors from the `valid-expect` rule as no runtime validation being called.

### Solution

Add an option `"allowTypeValidations": true` 

When this option is `true`, allow calls like `expect<T>(value: any)` when the generic type `T` explicitly provided.

**Comment by mrazauskas** [comment]
I think it would be better idea to use a type test runner instead. For example, [`jest-runner-tsd`](https://github.com/jest-community/jest-runner-tsd) could do the job.

Jest is meant to test JavaScript. `tsc` is meant to typecheck your code. None of them were made to testing types. In the example above the biggest issue is `@ts-expect-error` comment. For example, running the following through `tsc`:

```ts
expect<Foo>({ 
  // @ts-expect-error `bar` should be number
  br: 'string',
});
```

would pass. The comment would silence any syntax error as well. As you see that isn’t a good test.

`jest-runner-tsd` runs test using [`tsd`](https://github.com/SamVerschueren/tsd) which has `expectError()` assertion. That is not yet perfect, but somewhat better idea. 

---

By the way, `jest-runner-tsd` currently is running some 1000 type tests in Jest repo.

**Comment by G-Rath** [comment]
That is an interesting use of `expect` - I don't think its super terrible, but as @mrazauskas said there are strictly better tools for this  and it's really not how `expect` is meant to be used (e.g. this does not count as an assertion for `expect.hasAssertions`); you also should be able to handle this in userland by just creating a `expectType` utility:

```
export const expectType = <T>(value: T) => {};

type Foo = { bar: number };

expectType<Foo>({ bar: 123 });

expectType<Foo>({ 
  // @ts-expect-error `bar` should be number
  bar: 'string',
});
```

or by implementing a no-op `toMatchType` matcher:

```
expect.extend({
  toMatchType: () => ({ message: '', pass: true }),
});

type Foo = { bar: number };

expectType<Foo>({ bar: 123 }).toMatchType();

expectType<Foo>({ 
  // @ts-expect-error `bar` should be number
  bar: 'string',
}).toMatchType();
```

Given this, and that it's not come up before, I don't think it makes much sense for us to support.

**Comment by moshest** [comment]
Make sense. Thanks for the feedback!
